### PR TITLE
 Add analyzers to ref-pack / ASP.NET transport package

### PR DESCRIPTION
**Pull request by ericstj**

### CODE REVIEW

File: src/libraries/Directory.Build.props [hunk 1]
```diff
@@ -128,8 +128,9 @@
     <TestArchiveTestsDir>$(TestArchiveTestsRoot)$(OSPlatformConfig)/</TestArchiveTestsDir>
     <TestArchiveRuntimeRoot>$(TestArchiveRoot)runtime/</TestArchiveRuntimeRoot>
 
-    <ASPNETCoreAppPackageRefPath>$(ArtifactsBinDir)pkg\aspnetcoreapp\ref</ASPNETCoreAppPackageRefPath>
-    <ASPNETCoreAppPackageRuntimePath>$(ArtifactsBinDir)pkg\aspnetcoreapp\lib</ASPNETCoreAppPackageRuntimePath>
+    <ASPNETCoreAppPackageRootPath>$(ArtifactsBinDir)pkg\aspnetcoreapp\</ASPNETCoreAppPackageRootPath>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props:131]:
```suggestion
    <ASPNETCoreAppPackageRootPath>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'pkg', 'aspnetcoreapp'))</ASPNETCoreAppPackageRootPath>
```

File: src/libraries/Directory.Build.props [hunk 2]
```diff
@@ -128,8 +128,9 @@
...
+    <ASPNETCoreAppPackageRefPath>$(ASPNETCoreAppPackageRootPath)ref</ASPNETCoreAppPackageRefPath>
+    <ASPNETCoreAppPackageRuntimePath>$(ASPNETCoreAppPackageRootPath)lib</ASPNETCoreAppPackageRuntimePath>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props:133]: nit: we usually make sure that paths that point to directories end with a slash by using the NormalizeDirectory functions.

File: src/libraries/Directory.Build.targets [hunk 3]
```diff
@@ -42,6 +42,13 @@
     <IsNETCoreAppRef Condition="'$(IsNETCoreAppRef)' == '' and
                                 '$(IsReferenceAssembly)' == 'true' and
                                 $(NetCoreAppLibrary.Contains('$(AssemblyName);'))">true</IsNETCoreAppRef>
+
+    <AnalyzerForAssembly Condition="'$(AnalyzerForAssembly)' == '' AND ('$(IsGeneratorProject)' == 'true' or '$(IsAnalyzerProject)' == 'true')">$(AssemblyName.Substring(0, $(AssemblyName.LastIndexOf('.'))))</AnalyzerForAssembly>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:46]: nit: we usually use lower case operators:

```suggestion
    <AnalyzerForAssembly Condition="'$(AnalyzerForAssembly)' == '' and ('$(IsGeneratorProject)' == 'true' or '$(IsAnalyzerProject)' == 'true')">$(AssemblyName.Substring(0, $(AssemblyName.LastIndexOf('.'))))</AnalyzerForAssembly>
```

Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:46]: I was wondering if you couldn't use the GetFilePathWithoutExtension function here to strip the "extension" (which in this case of course isn't a file extension) away?

File: src/libraries/Directory.Build.targets [hunk 4]
```diff
@@ -97,6 +105,10 @@
       <RuntimePath>$(ASPNETCoreAppPackageRuntimePath)</RuntimePath>
       <RefPath>$(ASPNETCoreAppPackageRefPath)</RefPath>
     </BinPlaceTargetFrameworks>
+    <BinPlaceDir Include="$(ASPNETCoreAppPackageRootPath)\analyzers\dotnet\$(AnalyzerLanguage)"
+                 Condition="'$(IsAspNetCoreAppAnalyzer)' == 'true'" />
+    <BinPlaceDir Include="$(MicrosoftNetCoreAppRefPackDir)analyzers\dotnet\$(AnalyzerLanguage)"
+                 Condition="'$(IsNETCoreAppAnalyzer)' == 'true'" />
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:111]: Why not add support for analyzers in the binplacing targets similar to how I added support for binplacing native assets via `BinPlaceNative`: https://github.com/dotnet/arcade/blob/6b9d24236d8d1906284e6cb6c28e3fe93a69b7d2/src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/BinPlace.targets#L135?

File: src/libraries/Directory.Build.targets [hunk 5]
```diff
@@ -64,6 +71,7 @@
 
     <BinPlaceRef Condition="'$(BinPlaceRef)' == '' and ('$(IsReferenceAssembly)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRef>
     <BinPlaceRuntime Condition="'$(BinPlaceRuntime)' == '' and ('$(IsRuntimeAssembly)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRuntime>
+    <EnableBinplacing Condition="'$(IsAspNetCoreAppAnalyzer)' == 'true' or '$(IsNetCoreAppAnalyzer)' == 'true'">true</EnableBinplacing>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:74]: If you would add support for binplacing analyzers as suggested below you would also not need to set this property manually as binplacing would take care of it here: https://github.com/dotnet/arcade/blob/6b9d24236d8d1906284e6cb6c28e3fe93a69b7d2/src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/BinPlace.targets#L6

File: src/libraries/Directory.Build.targets (hunk 4 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:111]: We don’t need the TargetFramework information at all so we bypass that selection and just leverage the copy functionality.

File: src/libraries/Directory.Build.targets (hunk 5 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:74]: We would need some property like the above two, I’m not seeing that as any more beneficial. We don’t need a “proper name” support here. We can leverage binplacing generically.

File: src/libraries/Directory.Build.targets (hunk 3 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:46]: Sure I can do that. What’s your opinion about this vs some other convention? We could find the directory under libraries and use that for lookup.  We could have a direct property for analyzers and avoid the lookup.
I’m thinking the latter might be simpler.

File: src/libraries/Directory.Build.targets (hunk 5 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:74]: I could remove this property entirely by modifying this condition to also handle `'@(BinPlaceDir)' != ''` https://github.com/dotnet/arcade/blob/6b9d24236d8d1906284e6cb6c28e3fe93a69b7d2/src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/BinPlace.targets#L12

Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:74]: Yes that would be probably a better solution. I don't really like us setting `EnableBinplacing` here.

File: src/libraries/Directory.Build.targets (hunk 3 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:46]: Whatever you prefer. I'm fine with either the convention or with the inlined property.

Comment (ericstj) [reply src/libraries/Directory.Build.targets:46]: I think I prefer the simpler convention of `IsNETCoreAppAnalyzer` and `IsAspNetCoreAppAnalyzer` in the projects themselves.  Its less magic.

File: src/libraries/Directory.Build.targets (hunk 4 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:111]: https://github.com/dotnet/arcade/pull/7578

Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:106]: Why do code generators target netstandard2.0 and not NetCoreAppCurrent?

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.csproj [hunk 6]
```diff
@@ -7,6 +7,8 @@
     <NoWarn>CS1574</NoWarn>
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
     <UsingToolXliff>true</UsingToolXliff>
+    <IsNETCoreAppAnalyzer>true</IsNETCoreAppAnalyzer>
+    <AnalyzerLanguage>cs</AnalyzerLanguage>
```
Comment (safern) [inline src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.csproj:11]: Should we default `AnalyzerLanguage` to cs when `IsNETCoreAppAnalyzer` or `IsAspNetCoreAPpAnalyzer` are set to true and `AnalyzerLanguage` is empty?

**Review by safern — APPROVED** [review]

File: src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.csproj (hunk 6 above)
Comment (ericstj) [reply src/libraries/System.Text.Json/gen/System.Text.Json.SourceGeneration.csproj:11]: I'd prefer not: https://github.com/dotnet/runtime/pull/52554#discussion_r632060892.
I'd also like folks to force thinking about the fact that they chose to only support `cs`

File: src/libraries/Directory.Build.targets (hunk 4 above)
Comment (ericstj) [reply src/libraries/Directory.Build.targets:104]: Because they run in the compiler process which runs on both .NETFramework and .NETCore.

**Comment by ericstj** [comment]
Arm64 failure is https://github.com/dotnet/runtime/issues/54948

**Pull request merged**